File: src/main/java/com/lightning323/packInstaller/Main.java
package com.lightning323.packInstaller;

import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.dataformat.toml.TomlMapper;
import com.lightning323.packInstaller.fileTypes.FileEntry;
import com.lightning323.packInstaller.fileTypes.IndexFile;
import com.lightning323.packInstaller.fileTypes.PackConfig;
import com.lightning323.packInstaller.utils.IOUtils;

import java.io.BufferedReader;
import java.io.File;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.stream.Collectors;

import static com.lightning323.packInstaller.utils.IOUtils.fetchString;
import static com.lightning323.packInstaller.utils.IOUtils.getRelativeUrl;

public class Main {
    public static void main(String[] args) throws MalformedURLException {
        URL packTomlURL = new URL("https://raw.githubusercontent.com/Lightning323/MC-Terranova/refs/heads/main/pack/pack.toml");

        // Setup Mapper
        TomlMapper mapper = new TomlMapper();
        mapper.setPropertyNamingStrategy(PropertyNamingStrategies.KEBAB_CASE);

        try {
            System.out.println("Fetching pack configuration...");
            String packContent = fetchString(packTomlURL);

            // Deserialize PackConfig
            PackConfig config = mapper.readValue(packContent, PackConfig.class);

            System.out.println("--- Pack Info ---");
            System.out.println("Name: " + config.name);
            if (config.versions != null) {
                System.out.println("Minecraft Version: " + config.versions.get("minecraft"));
            }

            if (config.index != null) {
                System.out.println("\n--- Index ---");
                System.out.println("Index File Path: " + config.index.file);
                System.out.println("Hash Format: " + config.index.hashFormat);
                System.out.println("Hash: " + config.index.hash);

                //Get the index.toml
                URL indexURL = getRelativeUrl(packTomlURL, config.index.file);
                String indexContent = fetchString(indexURL);
                IndexFile indexData = mapper.readValue(indexContent, IndexFile.class);

                File saveDir = new File("./test_save/");
                saveDir.mkdirs();

                int i = 0;
                for (FileEntry entry : indexData.files) {
                    i++;
                    processFile(indexURL, saveDir, entry, i, indexData.files.size());
                }
            } else {
                System.err.println("No index found!");
            }

        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
        }
    }

    private static void processFile(URL baseURL, File saveDir, FileEntry entry, int index, int totalFiles) throws Exception {
        System.out.println("File (" + index + " of " + totalFiles + "): " + entry.file());
        IOUtils.downloadFile(baseURL, saveDir, entry);
    }


}

File: src/main/kotlin/com/lightning323/packInstaller/Main.kt
package lightning323.packInstaller.com.lightning323.packInstaller

import com.fasterxml.jackson.databind.PropertyNamingStrategies
import com.fasterxml.jackson.dataformat.toml.TomlMapper
import com.fasterxml.jackson.module.kotlin.registerKotlinModule
import java.net.HttpURLConnection
import java.net.URL

/**
 * Data structures matching your TOML format.
 * Using Kebab-Case strategy handles 'pack-format' and 'hash-format' automatically.
 */
data class PackConfig(
    val name: String = "",
    val author: String = "",
    val version: String = "",
    val packFormat: String = "",
    val index: IndexSection? = null,
    val versions: Map<String, String>? = null
)

data class IndexSection(
    val file: String,
    val hashFormat: String,
    val hash: String
)

/**
 * Helper to get HTTP response as String
 */
fun fetchString(url: String): String {
    val connection = URL(url).openConnection() as HttpURLConnection
    return connection.inputStream.use { stream ->
        stream.bufferedReader().readText()
    }
}

/**
 * Helper to get HTTP response as Binary
 */
fun fetchBinary(url: String): ByteArray {
    val connection = URL(url).openConnection() as HttpURLConnection
    return connection.inputStream.use { it.readBytes() }
}

fun main() {
    val packTomlURL = "https://raw.githubusercontent.com/Lightning323/MC-Terranova/refs/heads/main/pack/pack.toml"

    val mapper = TomlMapper().apply {
        registerKotlinModule()
        // Crucial for mapping 'pack-format' -> 'packFormat'
        propertyNamingStrategy = PropertyNamingStrategies.KEBAB_CASE
    }

    try {
        val result = fetchString(packTomlURL)

        // Deserialize into our object model
        val config = mapper.readValue(result, PackConfig::class.java)

        println("--- Pack Info ---")
        println("Name: ${config.name}")
        println("Minecraft Version: ${config.versions?.get("minecraft")}")

        config.index?.let { index ->
            println("\n--- Index ---")
            println("File: ${index.file}")
            println("Hash: ${index.hash}")
        }

    } catch (e: Exception) {
        println("Failed to parse TOML: ${e.message}")
    }
}